add article "Instant page navigations with speculation rules" (#55)
Zuehlke/software-engineering-corner: drafts/instant-page-navigation-with-speculation-rules.md

click on the link. This makes pre-rendering a potential vector for exploitation.

# Demonstration

I’ve created
a [demo blog site](https://speculation-rules-demo.deno.dev/) ([GitHub](https://github.com/vilben/speculation-rules-demo))
that includes a _landing_ page, an _about_ page, and a _blog_ page where all the posts are
displayed. The site also uses simple speculation rules to demonstrate how pre-rendering works:

```js
<script id="speculationrules" type="speculationrules">
    {
        "prerender": [{
        "source": "document",
        "where": {
        "and": [
    {"href_matches": "/*"},
    {"not": {"href_matches": "*-no-prerender*"}}
        ]},
        "eagerness": "eager"
    }]
    }
</script>
```

**Note**:
Currently, the blog page pre-renders all posts that match the speculation rules when it loads. In a
real-world scenario, you'd likely want to adjust the **eagerness** setting to balance resource usage
and performance.

## Eagerness

The **eagerness** setting determines when speculation rules are triggered.

- An eagerness of **"eager"** means that as soon as the page loads, it pre-renders every page that
  matches the speculation rules.

> - An eagerness of **"moderate"** delays prerendering until a user interacts with the

Reviewer: ```suggestion
- An eagerness of **"moderate"** delays pre-rendering until a user interacts with the
```

  page—specifically, when hovering over a link that meets the speculation rules criteria.

This setting helps balance resource usage and performance based on user interaction.
[More about eagerness](https://developer.chrome.com/docs/web-platform/prerender-pages#eagerness)

To see this in action, ensure you're using a supported browser and that your browser's "Preload
pages" setting is not turned off (unfortunately some extensions interfere with this setting), or
just use an incognito tab.

## Steps to test:

1) Navigate to the [demonstration blog site]( https://speculation-rules-demo.deno.dev/)
2) Open dev tools (F12)

Open the **Application** tab and find the *Background services* section ⇾ *Speculative Loads* ⇾
*Rules*, where you can view the active speculation rules.

Right next to *Rules* there's the *Speculations* section, you’ll see all the pages that have been

> pre-rendered or what state they are in. As they are in *Ready* state, When navigating, the page will

Reviewer: ```suggestion
pre-rendered or what state they are in. As they are in *Ready* state, when navigating, the page will
```

load instantly.
![Dev tools Application](https://cdn.hashnode.com/res/hashnode/image/upload/v1727713720845/jW-6gGIxw.png?auto=format)

You can also observe background activity in the **Network** tab, where you’ll notice the

> prerendering process at

Reviewer: ```suggestion
pre-rendering process at
```

work.![Dev tools Network](https://cdn.hashnode.com/res/hashnode/image/upload/v1727713751485/HsTee5a3T.png?auto=format)

## Excluded Links

Some links are intentionally excluded from pre-rendering. These are identified by URLs ending with "

> -no-prerender" and are also labeled with *"NO PRERENDER:"* in the link text for easy identification.

Reviewer: ```suggestion
-no-prerender" and are also labelled with *"NO PRERENDER:"* in the link text for easy identification.
```

## Performance Stats

In the top-left corner of each page, you'll find a *"Predictive Pre-rendering Stats"* box displaying
the page's load time (measured by Largest Contentful Paint, or LCP). While not 100% accurate, it’s
close enough for demonstration purposes.

By clicking through blog posts and checking the LCP time, you may not notice a speed improvement if
you're on a fast network. If that’s the case, I recommend throttling the network speed in DevTools:

1. In the **Network** tab, change the dropdown from *No throttling* to *Slow 4G* or even slower.
2. Test the pages again and compare load times.

> If the performance boost isn’t immediately obvious, it might be because the prerendering wasn't

Reviewer: ```suggestion
If the performance boost isn’t immediately obvious, it might be because the pre-rendering wasn't
```

> completed when the link was clicked. You can confirm pre-render status in the **Speculations**

Reviewer: ```suggestion
completed when the link was clicked. You can confirm the pre-render status in the **Speculations**
```

section under the **Application** tab.

In my tests, switching to *Slow 4G* resulted in about 2000ms for non-pre-rendered pages versus 100ms
for pre-rendered ones.

![Comparison "many images" slow 4G](https://cdn.hashnode.com/res/hashnode/image/upload/v1727713780823/5PK-RWGOS.gif?auto=format)

# Conclusion

In summary, the **Speculation Rules API** offers exciting opportunities to enhance the performance

> of Multi-Page Applications (MPAs) by intelligently pre-rendering or prefetching content, making

Reviewer: ```suggestion
of Multi-Page Applications (MPAs) by intelligently pre-rendering or pre-fetching content, making
```
